File: coin-change.cpp
class Solution
{
public:
    // TLE

    //     int coinCount(vector<int>& coins, int amount, int size){
    //         if(amount == 0 && size >= 0) return 0;
    //         if(amount != 0 && size == 0) return INT_MAX - amount;
    //         // int coin = coins[size-1];
    //         if(coins[size-1] > amount) return coinCount(coins, amount, size-1);
    //         else return min(coinCount(coins, amount, size - 1), 1+coinCount(coins, amount - coins[size-1], size));

    //     }
    //     int coinChange(vector<int>& coins, int amount) {
    //         int output = (coinCount(coins, amount, coins.size()));
    //         return output > amount ? -1 : output ;
    //     }

    int coinChange(vector<int> &coins, int amount)
    {
        int array[++amount];
        array[0] = 0;
        sort(coins.begin(), coins.end());
        for (int i = 1; i < amount; i++)
        {
            array[i] = INT_MAX;
            for (const int &j : coins)
            {
                if (i - j < 0)
                    break;
                if (array[i - j] != INT_MAX)
                    array[i] = min(array[i], 1 + array[i - j]);
            }
        }
        return array[--amount] == INT_MAX ? -1 : array[amount];
    }
};

File: coin_change.py
class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        # base case
        if amount == 0:
            return 0
        # initialized to large integers since we are taking minimums later
        dp = [10**9] * (amount)
        # for all coins values, we know they can be made with a single coin
        for coin in coins:
            if coin <= amount:
                dp[coin-1] = 1
        # for each amount value, we get the minimum number of coins 
        # needed to make it up by comparing it with all the previous steps given a 
        # coin to be added. 
        for i in range(amount):
            for c in coins:
                if i-c >=0:
                    dp[i] = min(dp[i], 1 + dp[i-c])
        # If not found we return -1
        return -1 if dp[-1] == 10**9 else dp[-1]